taichi/backends/opengl/opengl_api.cpp

```diff
@@ -373,13 +395,21 @@ struct CompiledProgram::Impl {
           ctx.args[i] = accum_size;
           accum_size += size;
         }  // concat all extptr into my baseptr
-        TI_INFO("baseptr = {}", baseptr);
         iov.push_back(IOV{baseptr, accum_size});
       }
     }
-    for (const auto &ker : kernels) {
+    {
       auto guard = launcher->create_launch_guard(iov);
-      ker->launch();
+      for (const auto &ker : kernels) {
+        if (ker->rse.has_value()) {
+          auto gtm_now = guard.map_buffer(1); // TODO: RAII
+          ker->num_groups = (*ker->rse)((const char *)gtm_now);
+          guard.unmap_buffer(1);
+        }
+        ker->launch();
+        glMemoryBarrier(GL_SHADER_STORAGE_BARRIER_BIT);
```

Comment:
TODO: Right now every OpenGL kernel launch will invoke `glMemoryBarrier`. On the other backends, they only sync when we have to read the GPU memory back to host. But I think doing this as a start is good enough..

Reply:
Actually this is only needed for NV GL's...

Reply:
ack